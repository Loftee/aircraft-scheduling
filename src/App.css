.app {
    width:100%;
    max-width:1200px;
    margin: 0 auto;
}

h3 {
    margin: 0 0 15px;
}