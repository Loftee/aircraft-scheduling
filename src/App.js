import React, {Component} from 'react';
import './App.css';
import AircraftList from './components/AircraftList';
import FlightList from "./components/FlightList";
import * as constants from './constants';
import * as services from './services';

class App extends Component {
    constructor() {
        super();
        this.state = {
            aircraftList: [],
            flightList: [],
            filteredFlightsList: []
        };
    }

    componentDidMount() {
        services.fetchAircrafts()
            .then(fetchedAircraft => {
                this.setState(prevState => ({
                    ...prevState,
                    aircraftList: fetchedAircraft
                }));
            });

        services.fetchFlights()
            .then(fetchedFlights => {
                this.setState(prevState => ({
                    ...prevState,
                    flightList: fetchedFlights
                }));
            });
    }

    selectAircraft = id => {
        const aircraftListCopy = JSON.parse(JSON.stringify(this.state.aircraftList));
        // Reset the selected aircraft list so multiple aircraft can't be selected at once.
        this.resetSelected(aircraftListCopy);
        const aircraftIndex = aircraftListCopy.findIndex(aircraft => aircraft.id === id);

        aircraftListCopy[aircraftIndex].selected = true;

        // Was going to have the flights filter when an aircraft was selected but
        // no flights match the aircraft's base.
        // const aircraftBase = aircraftListCopy[aircraftIndex].base;
        // this.filterFlights(aircraftBase, initialTime);

        // Set newly selected aircraft and reset flight list
        this.setState(prevState => ({
            ...prevState,
            aircraftList: aircraftListCopy,
            filteredFlightsList: this.state.flightList
        }));
    };

    selectFlight = (id) => {
        // Get the selected flight
        const flightListCopy = JSON.parse(JSON.stringify(this.state.flightList));
        const flightIndex = flightListCopy.findIndex(flight => flight.id === id);

        // Get flights arrival time and airport
        const destinationAirport = flightListCopy[flightIndex].destination;
        const arrivalTime = flightListCopy[flightIndex].arrivalTime + constants.TURN_AROUND_TIME;

        // Filter flights
        const filteredFlightsList = this.filterFlights(destinationAirport, arrivalTime);

        // Set filtered flights
        this.setState(prevState => ({
            ...prevState,
            filteredFlightsList: filteredFlightsList
        }));
    };

    resetSelected = (listCopy) => {
        listCopy.forEach(listItem => {
           listItem.selected = false
        });
    };

    filterFlights = (departAirport, departTime = 0) => {
        const flightListCopy = JSON.parse(JSON.stringify(this.state.flightList));
        const filteredByAirport = flightListCopy.filter(flight => flight.origin === departAirport);
        return filteredByAirport.filter(flight => flight.departureTime > departTime);
    };

    resetFlights = () => {
        this.setState(prevState => ({
            ...prevState,
            filteredFlightsList: this.state.flightList
        }));
    };

    render() {
        return (
            <div className={"app"}>
                <AircraftList {...this.state} selectAircraft={this.selectAircraft} />
                <FlightList {...this.state} selectFlight={this.selectFlight} resetFlights={this.resetFlights} />
            </div>
        );
    }
}

export default App;
