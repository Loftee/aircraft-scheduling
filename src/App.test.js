import React from 'react';
import {shallow} from "enzyme";
import App from './App';

describe('App', () => {
  const app = shallow(<App/>);

  it('renders correclty', () => {
    expect(app).toMatchSnapshot();
  });

  it('contains a AircraftList component', () => {
    expect(app.find('AircraftList')).toBeTruthy();
  });
}