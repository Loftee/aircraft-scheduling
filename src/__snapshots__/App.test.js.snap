// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="app"
>
  <AircraftList
    aircraftList={Array []}
    filteredFlightsList={Array []}
    flightList={Array []}
    selectAircraft={[Function]}
  />
  <FlightList
    aircraftList={Array []}
    filteredFlightsList={Array []}
    flightList={Array []}
    resetFlights={[Function]}
    selectFlight={[Function]}
  />
</div>
`;
